common/tier4_perception_rviz_plugin/src/tools/pedestrian_pose.cpp

```diff
@@ -47,25 +47,194 @@
 
 #include "pedestrian_pose.hpp"
 
-#include <rviz_common/display_context.hpp>
-#include <rviz_common/properties/float_property.hpp>
-#include <rviz_common/properties/string_property.hpp>
-
-#include <unique_identifier_msgs/msg/uuid.hpp>
+#include "util.hpp"
 
-#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
-#include <tf2_ros/transform_listener.h>
+#include <rviz_common/display_context.hpp>
 
 #include <algorithm>
+#include <limits>
+#include <memory>
 #include <random>
 #include <string>
 
 namespace rviz_plugins
 {
+InteractivePedestrian::InteractivePedestrian(const Ogre::Vector3 & point)
+{
+  velocity_ = Ogre::Vector3::ZERO;
+  point_ = point;
+  theta_ = 0.0;
+
+  std::mt19937 gen(std::random_device{}());
+  std::independent_bits_engine<std::mt19937, 8, uint8_t> bit_eng(gen);
+  std::generate(uuid_.begin(), uuid_.end(), bit_eng);
+}
+
+std::array<uint8_t, 16> InteractivePedestrian::uuid() const { return uuid_; }
+
+void InteractivePedestrian::twist(geometry_msgs::msg::Twist & twist) const
+{
+  twist.linear.x = velocity_.length();
+  twist.linear.y = 0.0;
+  twist.linear.z = 0.0;
+}
+
+void InteractivePedestrian::transform(tf2::Transform & tf_map2object) const
+{
+  tf2::Transform tf_object_origin2moved_object;
+  tf2::Transform tf_map2object_origin;
+
+  {
+    geometry_msgs::msg::Transform ros_object_origin2moved_object;
+    tf2::Quaternion quat;
+    quat.setRPY(0.0, 0.0, theta_);
+    ros_object_origin2moved_object.rotation = tf2::toMsg(quat);
+    tf2::fromMsg(ros_object_origin2moved_object, tf_object_origin2moved_object);
+  }
+
+  {
+    geometry_msgs::msg::Transform ros_object_map2object_origin;
+    tf2::Quaternion quat;
+    quat.setRPY(0.0, 0.0, 0.0);
+    ros_object_map2object_origin.rotation = tf2::toMsg(quat);
+    ros_object_map2object_origin.translation.x = point_.x;
+    ros_object_map2object_origin.translation.y = point_.y;
+    ros_object_map2object_origin.translation.z = point_.z;
+    tf2::fromMsg(ros_object_map2object_origin, tf_map2object_origin);
+  }
+
+  tf_map2object = tf_map2object_origin * tf_object_origin2moved_object;
+}
+
+void InteractivePedestrian::update(const Ogre::Vector3 & point)
+{
+  velocity_ = point - point_;
+  point_ = point;
+  theta_ = velocity_.length() < std::numeric_limits<double>::epsilon()
```

Comment:
I don't recommend using machine epsilon for floating value comparisons. Generally, the error of the calculation result will be larger.

In this case, the point has three axes, so if each has an epsilon error but represents the same position, the length of velocity will be root 3 epsilon.

Moreover, length uses xyz, but atan2 does xy only. So atan2(0, 0) is called if velocity_ is (0, 0, epsilon)

Reply:
OK, I see :+1: In short, the following implementation is better than original one, right?

```
theta_ = (velocity_.x < 1e-3 && velocity_.y < 1e-3) ? theta_ : atan2(velocity_.y, velocity_.x);
```

Reply:
I think so. (But I don't know what value is suitable for epsilon. It's very difficult.)

Reply:
For now, I'll fix as above example.

https://github.com/autowarefoundation/autoware.universe/pull/534/commits/133f3f7726711f5ef8222d6b18186011deefa3f7